### DISCUSSION TITLE
Disable `@prisma/client did not initialize yet` error in prod.

### DISCUSSION THREAD

**Discussion by Innei** [prisma/prisma#20960] [Q&A] [answered]
## Problem

When I execute the compiled JS file in the production environment, I get this error.

`@prisma/client did not initialize yet. Please run "prisma generate" and try to import it again`

The type files generated by Prisma are not needed in the production environment because using TypeScript and type check.

Even without type files, there should be no error here.

## Suggested solution

Even if there is no typing file here, it should not throw an error.

Disable this error reporting.

## Alternatives

 N/A

**Comment by aqrln** [comment]
If you get this error, the problem is not the missing `.d.ts` file (that indeed is not necessary in production) but the missing generated client itself (i.e., the generated JS file). Can you share more details about your setup and build steps?

**Reply by aqrln** [reply]
I just saw the other issue opened by you (https://github.com/prisma/prisma/issues/20948), I assume that is the root cause?

**Reply by Innei** [reply]
Yes, that is my main issue. And this problem is derived from another problem. I found in the client code that `prisma.xxxx.find` or other methods are all done through a proxy, so I understand that these codes may not need generator implementation. The generator only needs to be used for generating `d.ts` files.

**Reply by Innei** [reply]
The problem I am currently encountering is how to correctly COPY the generated client files in Docker. I am using pnpm and the generated path is uncertain. However, I should not install TypeScript in the production container and generate it there. It should only be done in the builder.

**Reply by aqrln** [reply]
Is there a reason you need to copy just the generated client specifically and not the whole `node_modules` directory?

**Reply by Innei** [reply, marked as answer]
I already solved, like this:

```
FROM node:16-alpine as builder
WORKDIR /app
COPY . .
RUN apk add git make g++ alpine-sdk python3 py3-pip unzip
RUN npm i -g pnpm
RUN pnpm install
RUN npm run build

FROM node:16-alpine
RUN apk add zip unzip bash --no-cache
RUN npm i -g pnpm
WORKDIR /app
COPY --from=builder /app/dist dist

ENV NODE_ENV=production
COPY package.json ./
COPY pnpm-lock.yaml ./
COPY .npmrc ./
COPY prisma ./prisma/
COPY external ./external/

COPY --from=builder /app/node_modules node_modules
RUN rm -rf node_modules/.cache
RUN pnpm prune
ENV TZ=Asia/Shanghai
EXPOSE 3333

CMD ["npm", "start:prod"]
```

**Reply by amruthpillai** [reply]
@aqrln To have a smaller final built image size?

**Comment by zwelhtetyan** [comment]
>The problem I am currently encountering is how to correctly COPY the generated client files in Docker. I am using pnpm and the generated path is uncertain.

I had the same problem a few days ago, my solution is custom generated prisma-client folder.

```ts
generator client {
  provider = "prisma-client-js"

  /**
   * ---------------------------------------------------------------------------------------------------------
   * By default, it generates client inside
   * `node_modules/.pnpm/@prisma+client@[version]_prisma@[version]_typescript@[version]__typescript@[version]`
   * which is tricky to copy that exact path in docker production image for extensible versions
   * So, generate custom output folder and copy that folder in docker-production-image
   * ---------------------------------------------------------------------------------------------------------
   * Note: Need to import prisma-client things from `customFolder/client` instead of default path `@prisma/client`
   * ---------------------------------------------------------------------------------------------------------
   * Tip: Add import alias in `tsconfig.json` file
   * {
   * "~prisma/*": ["node_modules/.prisma/*"]
   * }
   * With that alias, it should be for example,
   * ✅ `import { PrismaClient } from '~prisma/client'`
   * ❌ `import { PrismaClient } from '@prisma/client'`
   * ---------------------------------------------------------------------------------------------------------
   */
  output = "../node_modules/.prisma/client"
}
```

And just copy the necessary folder in docker production image
```bash
# Copy Prisma schema
COPY --from=build /usr/src/app/prisma ./prisma 

# Copy Prisma generated client
COPY --from=build /usr/src/app/node_modules/.prisma ./node_modules/.prisma
```
